File: models/IADBModel.m
//
//  NrstModel.m
//  airportsdb
//
//  Created by Christopher Hobbs on 2/18/14.
//  Copyright (c) 2014 Toonsy Net. All rights reserved.
//

#import "IADBModel.h"
#import "Airport.h"
#import "IADBPersistence.h"

static IADBPersistence *_persistence = nil;

@implementation IADBModel

+(IADBPersistence *) persistence {
    if( !_persistence ) {
        _persistence = [[IADBPersistence alloc] initWithPath:[[NSBundle mainBundle] pathForResource:@"airportsdb" ofType:@"sqlite"]];
    }
    return _persistence;
}

+(void) setPersistencePath:(NSString *) path {
    _persistence = [[IADBPersistence alloc] initWithPath:[[NSBundle mainBundle] pathForResource:@"airportsdb" ofType:@"sqlite"]];
}

+(NSManagedObjectContext *) managedObjectContext {
    return [[self persistence] managedObjectContext];
}

+(NSInteger) countAll {
    NSError *error;
    NSFetchRequest *fetch = [NSFetchRequest fetchRequestWithEntityName:[self description]];
    NSUInteger count = [[IADBModel managedObjectContext] countForFetchRequest:fetch error:&error];
    NSAssert3(!error, @"Unhandled error counting in %s at line %d: %@", __FUNCTION__, __LINE__, [error localizedDescription]);
    return count;
}

+(Airport *) findByAirportId:(int32_t) airportId {
    
    NSManagedObjectContext *context = [IADBModel managedObjectContext];
    NSEntityDescription *entityDescription = [NSEntityDescription entityForName:[self description] inManagedObjectContext:context];
    NSFetchRequest *request = [[NSFetchRequest alloc] init];
    [request setEntity:entityDescription];
    
    // Set example predicate and sort orderings...
    NSPredicate *predicate = [NSPredicate predicateWithFormat:
                              @"(airportId = %d)",airportId];
    [request setPredicate:predicate];
    
    NSError *error;
    NSArray *airports = [context executeFetchRequest:request error:&error];
    
    NSAssert3(airports, @"Unhandled error removing file in %s at line %d: %@", __FUNCTION__, __LINE__, [error localizedDescription]);
    
    if (!airports || airports.count == 0) { return nil; }
    if (airports.count > 1 ) { NSLog(@"WARNING: more than 1 %@ found with id %d",[self description],airportId); }
    
    return airports[0];
}

+(NSArray *) findAllByAirportId:(int32_t) airportId {
    
    NSManagedObjectContext *context = [IADBModel managedObjectContext];
    NSEntityDescription *entityDescription = [NSEntityDescription entityForName:[self description] inManagedObjectContext:context];
    NSFetchRequest *request = [[NSFetchRequest alloc] init];
    [request setEntity:entityDescription];
    
    // Set example predicate and sort orderings...
    NSPredicate *predicate = [NSPredicate predicateWithFormat:
                              @"(airportId = %d)",airportId];
    [request setPredicate:predicate];
    
    NSError *error;
    NSArray *objects = [context executeFetchRequest:request error:&error];
    
    NSAssert3(objects, @"Unhandled error removing file in %s at line %d: %@", __FUNCTION__, __LINE__, [error localizedDescription]);
    
    return objects;
}

@end

File: models/IADBModel.swift
//
//  IADBModel.swift
//  iAirportsDB
//
//  Created by Christopher Hobbs on 8/17/16.
//  Copyright © 2016 Toonsy Net. All rights reserved.
//

import Foundation
import CoreData

class IADBModel: NSManagedObject {
    static var persistence: IADBPersistence = {
        return IADBPersistence(path: NSBundle.mainBundle().pathForResource("iAirportsDB", ofType: "sqlite")!)
    }()
    
    class func setPersistantStorePath(path: String) {
        self.persistence = IADBPersistence(path: path)
        print("Persistant store path: \(path)")
    }
    
    class func clearPersistence() {
        self.persistence.persistentStoreClear()
        self.persistence = IADBPersistence(path: self.persistence.persistentStorePath)
        print("Cleared local db")
    }
    
    class func managedObjectContext() -> NSManagedObjectContext {
        return self.persistence.managedObjectContext
    }
    
    class func countAll() -> Int {
        let fetch = NSFetchRequest(entityName: self.description())
        var error: NSError? = nil
        let count = IADBModel.managedObjectContext().countForFetchRequest(fetch, error: &error)
        assert(error == nil, "Unhandled error counting in \(#function) at line \(#line): \(error!.localizedDescription)")
        return count
    }
    
    func isBlank() -> Bool {
        return self.valueForKey("airportId") == nil
    }
    
    class func findAllByAirportId(airportId: Int32) -> [IADBModel] {
        let context: NSManagedObjectContext = IADBModel.managedObjectContext()
        let entityDescription: NSEntityDescription = NSEntityDescription.entityForName(self.description(), inManagedObjectContext: context)!
        let request: NSFetchRequest = NSFetchRequest()
        request.entity = entityDescription
        // Set example predicate and sort orderings...
        let predicate: NSPredicate = NSPredicate(format: "(airportId = %d)", airportId)
        request.predicate = predicate
        do {
            let objects = try context.executeFetchRequest(request)
            if let models = objects as? [IADBModel] {
                return models
            } else {
                print( "Invalid types in findAllByAirportId \(objects)")
            }
        }
        catch let error as NSError {
            NSLog( "Unhandled error removing file in \(#function) at line \(#line): \(error.localizedDescription)")
        }
        return []
    }
    
    required override init(entity: NSEntityDescription, insertIntoManagedObjectContext context: NSManagedObjectContext?) {
        super.init(entity: entity, insertIntoManagedObjectContext: context)
    }
    
    func setCsvValues( values: [String: String] ) {
        // override me please
    }
}